### ISSUE TITLE
sql: add option to keep all table descriptor leases active on all nodes

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#23510]
https://github.com/cockroachdb/cockroach/pull/19005 made progress towards preventing table lease acquisition from impacting tail latency. We now renew table leases when they are in use and begin getting too close (< 1 minute) to their expiration.

However, if a table is not used for more than a minute (`DefaultTableDescriptorLeaseRenewalTimeout`), the next request that needs it _may_ find that its old lease has expired. Further, if a table is not used for more than five minutes (`DefaultTableDescriptorLeaseDuration`), the next request _is guaranteed to_ find that its old lease has expired. In these cases, the next request to use the table will first need to pay the latency cost of acquiring a table lease before it can do anything else. This is both surprising and potentially concerning for users of Cockroach who have strict latency requirements. Further, the effect of this issue is more pronounced in a geo-distributed cluster where a lease acquisition requires [multiple round-trips](https://github.com/cockroachdb/cockroach/issues/23506) that may take a significant amount of time.

Poorly remembered quote from @bdarnell:
> "Back when using BigTable, I used to run a script to continually renew all leases. I never want to pay the latency cost of renewing a table lease synchronously."

We should add an option to automatically renew table leases continuously in the background even if they are not being used actively. This option could apply to all tables or to only a subset of tables. It would probably take the form of a `SESSION SETTING`, but it may make sense as a `CLUSTER SETTING` as well.

The downsides to this are: 
- lease acquisitions for a table may be wasted work if the table is never used 

cc. @andreimatei @vivekmenezes 

Jira issue: CRDB-5813

**Comment by andreimatei** [comment]
But they why "keep all table descs" rather than acquire all leases on server startup?

**Comment by nvanbenschoten** [comment]
> But then why "keep all table descs" rather than acquire all leases on server startup?

That's another option that's mostly equivalent. Either way, the important part is that we'd indefinitely renew table leases.

**Comment by petermattis** [comment]
I'm moving this to 2.1 as it doesn't seem very hard and the effect is being seen in the real world.

**Comment by awoods187** [comment]
@vivekmenezes @nstewart we need to figure out how to get this into 2.1 as @robert-s-lee says he keeps encountering this as a problem in pocs

**Comment by nstewart** [comment]
@mjibson can you take a look at this from a cost perspective? We can see how it compares to our other priorities for this milestone or the next after you've reviewed it.

**Comment by nstewart** [comment]
Spoke with the team. We are targeting this for 2.1

**Comment by dt** [comment]
An all/nothing setting seems like it'd be risky to default to on, as it could be a problem for a cluster with thousands of tables, but would be unlikely to benefit to majority of users if defaulted to off.

I'm thinking a simple option might be instead to just pre-lease the first X tables, with X controlled via a setting and has a default s.t. we'd expect it to be "all" for most clusters, but still avoid issues for clusters with very large numbers of tables.  Since we expect it to be "all tables" most of the time, this could be something simple, like just the first X in the descriptor table?

**Comment by robert-s-lee** [comment]
Being able to blacklist/whitelist the tables by names or regex would be helpful if all/nothing has to be done.

**Comment by mjibson** [comment]
Is a zone config setting an unreasonable suggestion?

**Comment by dt** [comment]
I went ahead and tried a simple proactive lease of 50 tables goroutine.

With my 4 node roachprod cluster, both before and after proactively acquiring leases, I still see a 120 -> 3ms speedup between the first and second query. I tried `SHOW KV TRACE` for them and indeed the trace is different the second time, with the first needing to do some extra steps like populate the distsender range cache it looks like:

```
root@localhost:26257/defaultdb> show kv trace for select baz.id, bar.id, foo.id from baz left outer join bar on baz.barid = bar.id left outer join foo
+----------------------------------+----------------+-------------------------------------------------------------+-----------------------------------------------------------------------------+-----+---------------+------+
|            timestamp             |      age       |                           message                           |                                     tag                                     | loc |   operation   | span |
+----------------------------------+----------------+-------------------------------------------------------------+-----------------------------------------------------------------------------+-----+---------------+------+
| 2018-06-26 22:13:00.967175+00:00 | 558µs803ns     | Scan /Table/54/{1-2}                                        | [n5,client=127.0.0.1:60474,user=root]                                       |     | starting plan |    1 |
| 2018-06-26 22:13:00.967306+00:00 | 690µs211ns     | querying next range at /Table/54/1                          | [client=127.0.0.1:60474,user=root,txn=7d209421,n5]                          |     | dist sender   |    2 |
| 2018-06-26 22:13:00.967394+00:00 | 778µs42ns      | querying next range at /Meta2/Table/54/1/NULL               | [client=127.0.0.1:60474,user=root,txn=7d209421,range-lookup=/Table/54/1,n5] |     | dist sender   |    4 |
| 2018-06-26 22:13:00.967444+00:00 | 828µs385ns     | r1: sending batch 1 Scan to (n3,s3):3                       | [client=127.0.0.1:60474,user=root,txn=7d209421,range-lookup=/Table/54/1,n5] |     | dist sender   |    4 |
| 2018-06-26 22:13:00.968778+00:00 | 2ms161µs651ns  | r33: sending batch 1 Scan to (n5,s5):5                      | [client=127.0.0.1:60474,user=root,txn=7d209421,n5]                          |     | dist sender   |    2 |
| 2018-06-26 22:13:00.969243+00:00 | 2ms626µs777ns  | querying next range at /System/NodeLiveness/5               | [s5,r33/5:/Table/5{4-5},txn=5a1d25a5,n5]                                    |     | dist sender   |   11 |
| 2018-06-26 22:13:00.969276+00:00 | 2ms660µs66ns   | r3: sending batch 1 CPut, 1 BeginTxn, 1 EndTxn to (n2,s2):2 | [s5,r33/5:/Table/5{4-5},txn=5a1d25a5,n5]                                    |     | dist sender   |   11 |
| 2018-06-26 22:13:01.03795+00:00  | 71ms333µs739ns | Scan /Table/53/{1-2}                                        | [n5,client=127.0.0.1:60474,user=root]                                       |     | starting plan |    1 |
| 2018-06-26 22:13:01.037975+00:00 | 71ms359µs59ns  | querying next range at /Table/53/1                          | [client=127.0.0.1:60474,user=root,txn=7d209421,n5]                          |     | dist sender   |   14 |
| 2018-06-26 22:13:01.038066+00:00 | 71ms449µs442ns | querying next range at /Meta2/Table/53/1/NULL               | [client=127.0.0.1:60474,user=root,txn=7d209421,range-lookup=/Table/53/1,n5] |     | dist sender   |   16 |
| 2018-06-26 22:13:01.038109+00:00 | 71ms492µs496ns | r1: sending batch 1 Scan to (n3,s3):3                       | [client=127.0.0.1:60474,user=root,txn=7d209421,range-lookup=/Table/53/1,n5] |     | dist sender   |   16 |
| 2018-06-26 22:13:01.039378+00:00 | 72ms761µs676ns | r32: sending batch 1 Scan to (n1,s1):4                      | [client=127.0.0.1:60474,user=root,txn=7d209421,n5]                          |     | dist sender   |   14 |
| 2018-06-26 22:13:01.040893+00:00 | 74ms276µs845ns | Scan /Table/55/{1-2}                                        | [n5,client=127.0.0.1:60474,user=root]                                       |     | sql txn       |    0 |
| 2018-06-26 22:13:01.04092+00:00  | 74ms303µs801ns | querying next range at /Table/55/1                          | [client=127.0.0.1:60474,user=root,txn=7d209421,n5]                          |     | dist sender   |   21 |
| 2018-06-26 22:13:01.04096+00:00  | 74ms344µs194ns | r34: sending batch 1 Scan to (n5,s5):5                      | [client=127.0.0.1:60474,user=root,txn=7d209421,n5]                          |     | dist sender   |   21 |
| 2018-06-26 22:13:01.041377+00:00 | 74ms760µs894ns | querying next range at /System/NodeLiveness/5               | [s5,r34/5:/{Table/55-Max},txn=1293aef5,n5]                                  |     | dist sender   |   27 |
| 2018-06-26 22:13:01.041414+00:00 | 74ms798µs121ns | r3: sending batch 1 CPut, 1 BeginTxn, 1 EndTxn to (n2,s2):2 | [s5,r34/5:/{Table/55-Max},txn=1293aef5,n5]                                  |     | dist sender   |   27 |
+----------------------------------+----------------+-------------------------------------------------------------+-----------------------------------------------------------------------------+-----+---------------+------+
(17 rows)

Time: 139.721075ms
```

And second:
```
root@localhost:26257/defaultdb> show kv trace for select baz.id, bar.id, foo.id from baz left outer join bar on baz.barid = bar.id left outer join foo on bar.fooid = foo.id;
+----------------------------------+---------------+----------------------------------------+----------------------------------------------------+-----+---------------+------+
|            timestamp             |      age      |                message                 |                        tag                         | loc |   operation   | span |
+----------------------------------+---------------+----------------------------------------+----------------------------------------------------+-----+---------------+------+
| 2018-06-26 22:13:08.702982+00:00 | 472µs913ns    | Scan /Table/54/{1-2}                   | [n5,client=127.0.0.1:60474,user=root]              |     | starting plan |    1 |
| 2018-06-26 22:13:08.703016+00:00 | 507µs101ns    | querying next range at /Table/54/1     | [client=127.0.0.1:60474,user=root,txn=0412269f,n5] |     | dist sender   |    2 |
| 2018-06-26 22:13:08.703056+00:00 | 546µs971ns    | r33: sending batch 1 Scan to (n5,s5):5 | [client=127.0.0.1:60474,user=root,txn=0412269f,n5] |     | dist sender   |    2 |
| 2018-06-26 22:13:08.703252+00:00 | 743µs624ns    | Scan /Table/53/{1-2}                   | [n5,client=127.0.0.1:60474,user=root]              |     | starting plan |    1 |
| 2018-06-26 22:13:08.703267+00:00 | 757µs854ns    | querying next range at /Table/53/1     | [client=127.0.0.1:60474,user=root,txn=0412269f,n5] |     | dist sender   |    4 |
| 2018-06-26 22:13:08.703289+00:00 | 780µs619ns    | r32: sending batch 1 Scan to (n3,s3):2 | [client=127.0.0.1:60474,user=root,txn=0412269f,n5] |     | dist sender   |    4 |
| 2018-06-26 22:13:08.704336+00:00 | 1ms826µs999ns | Scan /Table/55/{1-2}                   | [n5,client=127.0.0.1:60474,user=root]              |     | sql txn       |    0 |
| 2018-06-26 22:13:08.704373+00:00 | 1ms864µs807ns | querying next range at /Table/55/1     | [client=127.0.0.1:60474,user=root,txn=0412269f,n5] |     | dist sender   |    7 |
| 2018-06-26 22:13:08.704425+00:00 | 1ms916µs229ns | r34: sending batch 1 Scan to (n5,s5):5 | [client=127.0.0.1:60474,user=root,txn=0412269f,n5] |     | dist sender   |    7 |
+----------------------------------+---------------+----------------------------------------+----------------------------------------------------+-----+---------------+------+
(9 rows)

Time: 4.899299ms
```

So, I can clean up this proactive lease thing for a PR, but it doesn't look like it'll actually solve the problem on its own.

**Comment by awoods187** [comment]
@dt given the trace, any idea what is actually causing the delay? What do you think might solve the underlying problem?

**Comment by nvanbenschoten** [comment]
@dt there are a few major differences between priming the dist sender and priming table descriptor leases that is missing from the conclusion you're drawing. The first is that populating the range cache only requires reads while acquiring a table descriptor lease requires writing. This makes a big difference in distributed clusters. The second difference and the one that prompted this issue to be opened is that range descriptors don't expire but table leases do. It's tough to tell from the trace because I think it's missing important parts, but if you ran this same query again in 5 minutes you would be hit again with the latency of refreshing all relevant table leases. This is the specific problem that came up a number of times in POCs and when manual querying tables because the frequency of access to tables in these demos is often not enough to get them to automatically re-aquire table leases asynchronously.

That said, you bring up a very interesting separate point about priming dist sender level caches (rangeDescriptorCache, leaseholderCache). This is something we've discussed in the context of benchmarking, but never pursued. Ideally we would have a way to prime all of these caches and force ranges to establish leaseholders without needing to actually scan all of the data they contain.

**Comment by dt** [comment]
@nvanbenschoten I should have clarified that proactive leasing _does_ help, just not enough to say that it will eliminate the first query "warmup" effect.

I've been digging a little more here with a 4 node, single region mostly empty cluster. Proactive leasing *does* appear to have some effect -- without it my first query takes  ~130ms before dropping to 5ms on the second, while with it, my first query takes ~70ms.

**Comment by bdarnell** [comment]
> Ideally we would have a way to prime all of these caches and force ranges to establish leaseholders without needing to actually scan all of the data they contain.

The replica scanner should force ranges to establish leaseholders (over a 10 minute period), although it won't do anything to populate the caches.

**Comment by nvanbenschoten** [comment]
> just not enough to say that it will eliminate the first query "warmup" effect.

Absolutely, I was just trying to draw a distinction between the "warmup" effect and the "cooloff" effect as leases expire. Proactive leasing should eliminate the latter effect, which is what this issue is primarily focused on because its arguably a lot more surpring to users.

**Comment by awoods187** [comment]
So we should test the latencies after 5 minutes and 10 minutes with @dt's changes enabled to see if the "cooloff" effect is addressed. If so, we can leave with the initial startup improvement of 50%.

**Comment by dt** [comment]
@andreimatei was not a fan of my quick fix in #27017 so this is back to waiting on someone to find time to do a fancier fix.

**Comment by vivekmenezes** [comment]
@dt @andreimatei I don't see a strong reason for a query needing a descriptor to need to wait for a lease on it. I think a request can begin the process of acquiring a lease for future requests, but it can move forward with directly reading a descriptor from the store in the event that it can't find a descriptor in the cache. That would certainly have it not block on the write to the lease table (going through raft).

**Comment by asubiotto** [comment]
@lucy-zhang to triage/close

**Comment by postamar** [comment]
@ajwerner had explained to me that he wanted to replace the expiration column in `system.lease` with a `session_id` reference to sql liveness. I'm going to close this issue because whatever we end up doing, it's almost certainly not going to be what's described here. Please correct me and reopen if I'm wrong and/or completely off base here.

**Comment by postamar** [comment]
Here's the issue I was referring to above: https://github.com/cockroachdb/cockroach/issues/61419
